### DISCUSSION TITLE
korean language recognition

### DISCUSSION THREAD

**Discussion by khawar-islam** [mindee/doctr#974] [General]
I simply run your example code on Korean image it detect all the Korean characters but in your below question you didn't specify Korean language in your OCR. What is the reason?

https://github.com/mindee/doctr/discussions/837

**Code**

```
import glob
import os
os.environ['USE_TORCH'] = '1'
import cv2
import os
from doctr.io import DocumentFile
from doctr.models import ocr_predictor, crnn_mobilenet_v3_large

testing_path = "/media/cvpr/CM_1/COREMAX/testing/"
result_path = "/media/cvpr/CM_1/COREMAX/docTR/"

for img in glob.glob(testing_path + '*.*'):
    if img.endswith(".pdf"):
        doc = DocumentFile.from_pdf(img)
    else:
        doc = DocumentFile.from_images(img)

    img_bgr_rgb = cv2.imread(img)
    file_Name = os.path.basename(img)
    image = DocumentFile.from_images(img)
    model = ocr_predictor(det_arch='db_resnet50', reco_arch='crnn_mobilenet_v3_large', pretrained=True)
    result = model(image)
    export = result.export()
    result.show(doc)
```
Output
![Screenshot from 2022-07-04 13-37-50](https://user-images.githubusercontent.com/11488932/177082696-30247620-88f6-42ed-9d01-ea9e1e6669cb.png)

Now, I am applying OCR to classify Korean text and hide some confidential part in Korean document/image. At the moment, the OCR only detect the English characters and digits not hangul/korean characters. **Any way to fix the problem?**

**Code**

```
import glob
import os

os.environ['USE_TORCH'] = '1'
import cv2
import os
from doctr.io import DocumentFile
from doctr.models import ocr_predictor, crnn_mobilenet_v3_large

testing_path = "/media/cvpr/CM_1/COREMAX/testing/"
result_path = "/media/cvpr/CM_1/COREMAX/docTR/"

for img in glob.glob(testing_path + '*.*'):
    if img.endswith(".pdf"):
        doc = DocumentFile.from_pdf(img)
    else:
        doc = DocumentFile.from_images(img)

    img_bgr_rgb = cv2.imread(img)
    file_Name = os.path.basename(img)

    image = DocumentFile.from_images(img)

    model = ocr_predictor(det_arch='db_resnet50', reco_arch='crnn_mobilenet_v3_large', pretrained=True)
    result = model(image)

    export = result.export()

    #result.show(doc)
    print(export)

    page_words = [[word for block in page['blocks'] for line in block['lines'] for word in line['words']] for page
                  in export['pages']]

    page_dims = [page['dimensions'] for page in export['pages']]

    words_abs_coords = [
        [[int(round(word['geometry'][0][0] * dims[1])), int(round(word['geometry'][0][1] * dims[0])),
          int(round(word['geometry'][1][0] * dims[1])), int(round(word['geometry'][1][1] * dims[0]))] for word in words]
        for words, dims in zip(page_words, page_dims)
    ]

    for words, dims in zip(page_words, page_dims):
        for word in page_words:
            for single_word in word:
                if single_word['value'].__contains__('-'):
                    xmin = int(single_word['geometry'][0][0] * dims[1])
                    ymin = int(single_word['geometry'][0][1] * dims[0])
                    xmax = int(single_word['geometry'][1][0] * dims[1])
                    ymax = int(single_word['geometry'][1][1] * dims[0])
                    image = cv2.rectangle(img_bgr_rgb, (xmin, ymin), (xmax, ymax), (235, 87, 51), cv2.FILLED)
                    cv2.imwrite(os.path.join(result_path, file_Name), img_bgr_rgb)](url)

```

**Output**
![aaa](https://user-images.githubusercontent.com/11488932/177083431-5fefd488-d9a1-43b7-8694-fe4e5561cddb.jpg)

**Comment by felixdittrich92** [comment]
Hi @khawar-islam :wave:,

the reason is that the pretrained models are trained on the french vocab so it can only recognize characters which are inside this vocab.
the only way currently would be to add the korean vocab in [VOCABS](https://github.com/mindee/doctr/blob/main/doctr/datasets/vocabs.py) and train a recognition model on your own :)

You can find some details how it works and how you have to prepare your dataset in references : [reco train](https://github.com/mindee/doctr/tree/main/references/recognition)

**Reply by khawar-islam** [reply]
@felixdittrich92  thank you. Do you have any tutorial regarding how to prepare data?

**Reply by felixdittrich92** [reply]
Unfortunately no only the resources in references (we have planned some tutorials for v0.6.0).
You could try our WordGenerator (until korean vocab is added) but this needs some improvements .. for the moment i would suggest to use https://github.com/Belval/TextRecognitionDataGenerator for synthetic dataset creation :) The dataset structure is really easy i will attach a small example in this thread tomorrow

**Reply by felixdittrich92** [reply]
@khawar-islam 
I have attached a minor example :)
[sample.zip](https://github.com/mindee/doctr/files/9052173/sample.zip)

Some additional information:
- sar_resnet31 and master works only on main branch in both frameworks if you want to train any of these models (will be released in 0.5.2 soon)
- if latency is relevant and your text scences are regular (mostly in documents) you should use a recognition model which uses mobilenetv3 as backbone 
- as explained all pretrained models are trained on french vocab in any case of non latin vocabs i would suggest to train it from scratch (~10M split in train and val should be good)

**Reply by khawar-islam** [reply]
@felixdittrich92 I really want to train from scratch but I have limited resources like only one GPU with 12 GB. Can i train on it with 10M images? I think not.

**Reply by felixdittrich92** [reply]
Hi @khawar-islam 👋 ,
sure but you have to experiment a bit with the batch size to find a value which fits. 
And of course it depends on your hardware how long it takes as a comparable example:

My system:
RTX 3080 mobile (16GB) / 32 GB Ram
MjSynth 500k split takes ~25min

**Reply by khawar-islam** [reply]
@felixdittrich92 thank you. According to suggestion, I have generated 194,486 samples containing blur, back,basic, dis and skew images using https://github.com/Belval/TextRecognitionDataGenerator . But i do not have "labels.json" file. Do i have to generate separately?

![Screenshot from 2022-08-05 16-59-58](https://user-images.githubusercontent.com/11488932/183032720-7b945c2f-2692-41a1-9c0d-032b0730d101.png)

**Reply by frgfm** [reply]
Hi @khawar-islam :wave:

Yes you'll need to create it yourself, and you can find all the relevant details here :) 
https://github.com/mindee/doctr/tree/main/references/recognition#data-format
